[Dependency Scanning] Restrict Swift overlay lookup to "visible" Clang modules only

artemcm opened the pull request:
https://github.com/llvm/llvm-project/pull/147969 added support to the Clang dependency scanner to query the set of "visible" modules for a given TU or named query. Swift must utilize this data to driver lookup of Swift module overlays. Previously, the scanner implementation queried Swift module overlay dependencies for a Swift module by checking *all* of its imported Clang modules, direct and transitive - even when not re-exported. With this change, the scanner's overlay lookup is only done for a Swift module's "visible" set of Clang module dependencies as reported by the scanner, matching the behavior expected to occur in the subsequent compilation task.

artemcm committed:
[Dependency Scanner] Adapt to API change on Clang's 'DependencyScanningTool::getModuleDependencies'

artemcm committed:
[Dependency Scanning] Restrict Swift overlay lookup to "visible" Clang modules only

Previously Swift overlay lookup was performed for every directly and transitively-imported Clang module.

https://github.com/llvm/llvm-project/pull/147969 introduced the concept of "visible" Clang modules from a given named Clang dependency scanner query which closely maps to the set of modules for which Swift will attempt to load a Swift overlay. This change switches overlay querying to apply only to the set of such visible modules.

Resolves rdar://144797648

artemcm commented:
@swift-ci test

cyndyishida commented on lib/DependencyScan/ModuleDependencyScanner.cpp:
```diff
@@ -1473,33 +1466,26 @@ void ModuleDependencyScanner::resolveHeaderDependenciesForModule(
-    findAllImportedClangModules(dep.ModuleName, cache, allClangDependencies,
-                                knownModules);
-
+  auto visibleClangDependencies = cache.getVisibleClangModules(moduleID);
```
question regarding how the cache works with https://github.com/swiftlang/swift/blob/15593e5ca68b7e3fba358fe617fd38fc66bdf879/lib/DependencyScan/ModuleDependencyScanner.cpp#L1127 
and https://github.com/swiftlang/swift/blob/15593e5ca68b7e3fba358fe617fd38fc66bdf879/lib/DependencyScan/ModuleDependencyScanner.cpp#L1176

Consider a direct dependency clang module `Foo` that has a transitive dependency `Bar`. If I read this correctly, 
after getting `Foo`'s dependency info from clang, there would be a cache hit on `Bar`'s moduleID if it was later observed as a direct dependency. But would there be information about the visible modules from importing `Bar`? I suspect the correct behavior would be to ask the clang scanning service.

artemcm replied:
That's a great catch!
Yeah, since the visible module set is per-query and not per-module, I think we might have to query every single module directly imported from Swift, even if it's been previously reported as a transitive dependency. I'm gonna give some thought to how best to organize that and see if I can measure what perf impact this will have.

artemcm committed:
[Dependency Scanning] Ensure all direct Clang module imports from Swift modules are always queried by-name to the Clang dependency scanner

This is required in order to always have computed the set of visible Clang modules for each Swift module in the graph. Otherwise when some Clang module gets cached as a transitive dependency from a query and is later looked up as a direct dependency, there will not be any computed visible modules set.

artemcm commented:
https://github.com/swiftlang/llvm-project/pull/10987
@swift-ci test

cyndyishida reviewed (approved):
LGTM

artemcm commented:
https://github.com/swiftlang/llvm-project/pull/10987
@swift-ci test

artemcm commented:
https://github.com/swiftlang/llvm-project/pull/10987
@swift-ci test

artemcm commented:
https://github.com/swiftlang/llvm-project/pull/10987
@swift-ci test

artemcm commented:
https://github.com/swiftlang/llvm-project/pull/10987
@swift-ci test

artemcm commented:
https://github.com/swiftlang/llvm-project/pull/10987
@swift-ci test macOS platform

artemcm commented:
https://github.com/swiftlang/llvm-project/pull/10987
@swift-ci test Linux platform

artemcm commented:
https://github.com/swiftlang/llvm-project/pull/10987
@swift-ci test
